BUG: to_datetime with unit with Int64

AlexKirko opened the pull request:
- [X] closes #30050
- [X] tests added 1 / passed 1
- [X] passes `black pandas`
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

Test output:
```
$ py.test pandas/tests/tools/test_datetime.py
============================= test session starts =============================
platform win32 -- Python 3.7.3, pytest-5.3.1, py-1.8.0, pluggy-0.13.0
rootdir: C:\git_contrib\pandas\pandas, inifile: setup.cfg
plugins: hypothesis-4.50.6, cov-2.8.1, forked-1.1.2, xdist-1.30.0
collected 1 item

pandas\tests\tools\test_datetime.py .                                    [100%]

============================== 1 passed in 0.03s ==============================

```

Notes: Hello. The fix introduces one additional type check if the Series passed to `to_datetime` stores anything except and IntegerArray. It does nothing else in this case. If we try to pass an IntegerArray, it pulls all the values that aren't NaN into an ndarray, converts that into datetime and adds NaT in the places where NaNs were. No precision is lost.

AlexKirko committed:
BUG: close #30050 - initial solution

jreback commented on pandas/tests/tools/test_datetime.py:
```diff
@@ -0,0 +1,23 @@
+import numpy as np
+
```
don't add a new file, instead add a test in pandas/tests/indexes/datetimes/test_tools.py

AlexKirko replied:
Moved the test, thanks.

jreback commented on pandas/core/tools/datetimes.py:
```diff
@@ -316,7 +317,20 @@ def _convert_listlike_datetimes(
         if format is not None:
             raise ValueError("cannot specify both format and unit")
         arg = getattr(arg, "values", arg)
```
this is the right idea, but needs to be done more generically, w/o peering inside EAs.

jreback replied:
change this
```arg = getattr(arg, "values", arg)```
to 
```arg = getattr(arg, "_values", arg)```

which will extract the EAs w/o conversion.

then create a new function where you do the conversion

jreback replied:
have a branch that checks if its an EA in an if/else
then use *only* the public functions, e.g. ``arr.isna()`` (which returns the mask)
use ``arr.na_value`` (for the NA value)
and use ``.take()`` to re-assemble.

AlexKirko replied:
Hello. 

Sorry, I don't quite understand some of your suggestions. I made all the changes I knew how to do and tried to move from inner implementations to public functions. Specifically, I couldn't come up with a way to reassemble numpy arrays split by arbitrary indices using `.take()` , so I instead used `np.insert` to reassemble. This allowed me to stay away from specifying datatypes, preallocating an array, and assigning data by index.

Specific changes:
1.  Change `arg = getattr(arg, "values", arg)` to `arg = getattr(arg, "_values", arg)`. As far as I can tell, this makes no other changes necessary. Could you clarify?
2.  Switch from `arg._mask` to `arg.isna()`. Thanks.
3.  Switch to np.compress for grabbing the non-nan values from arg to generalize the code. Pass the result directly into `tslib.array_with_unit_to_datetime` to avoid storing it.
4.  Switch from preallocating an ndarray of datetime64 to using np.insert to inject NaT into the result.

Could you clear up what you meant by using `.take()` to re-assemble the result? I also avoided using `arr.na_value`, because I was only able to find it for pandas arrays.

Test results:
```
py.test pandas/tests/indexes/datetimes/test_tools.py -k 'test_intarray_to_datetime'
============================= test session starts =============================
platform win32 -- Python 3.7.3, pytest-5.3.1, py-1.8.0, pluggy-0.13.0
rootdir: D:\git_contrib\pandas\pandas, inifile: setup.cfg
plugins: hypothesis-4.53.2, cov-2.8.1, forked-1.1.2, xdist-1.30.0
collected 416 items / 415 deselected / 1 selected

pandas\tests\indexes\datetimes\test_tools.py .                           [100%]

====================== 1 passed, 415 deselected in 0.19s ======================
```

AlexKirko committed:
CLN: switch to generic functions

Stop relying on inner implementations of IntegerArray and numpy arrays.
Use public funcitons as much as possible.

jreback commented:
@AlexKirko the diff is super confusing because you are pulling on other commits, pls merge in master and repush

AlexKirko commented:
> @AlexKirko the diff is super confusing because you are pulling on other commits, pls merge in master and repush

@jreback Apologies for the mess. Still learning how to contribute Git properly. Merged master in.

AlexKirko commented:
@jreback Conflicts in whatsnew removed. Should be ready for another review.

AlexKirko commented:
@jreback Merged in master again to remove new conflicts in `datetime.py`

jreback commented on pandas/core/tools/datetimes.py:
```diff
@@ -316,8 +317,23 @@ def _convert_listlike_datetimes(
+        arg = getattr(arg, "_values", arg)
+        # GH 30050 pass an ndarray to tslib.array_with_unit_to_datetime
+        # because it expects an ndarray argument
+        if isinstance(arg, IntegerArray):
```
ok this soln looks ok. wonder if we pass the values AND the mask directly to ```tslib.array_with_unit_to_datetime``` (pass an np.zeros mask for the non-integer NA case). would make this simpler from a code perspective (should be the same from perf).

this way all of the nan handling is in the same place.

AlexKirko replied:
@jreback We'd need to add `mask` as an argument to `tslib.array_with_unit_to_datetime` (with ndarray type), then we'd fill it with zeros whenever a non-nullable int type is passed and introduce an `if` to check. 

However, in `datetimes.py` we'd need to get `_ndarray_values` from `IntegerArray` to pass the full ndarray of values to `tslib.array_with_unit_to_datetime` (which would rely on the implementation of non-public stuff in `IntegerArray`) or we'd need to pass the incomplete values array in `_convert_listlike_datetimes` and rebuild the full array in `tslib.array_with_unit_to_datetime`, which would differ from the approach taken with other nullable datatypes.

Do you see any other way to accomplish moving the nan-handling to `tslib.array_with_unit_to_datetime` or is one of these two approaches acceptable?

jreback replied:
i guess my comment was not clear

pass the values and mask to array_with_unit_to_datetime

AlexKirko commented:
Moved the nan-handling to tslib. Web_and_Docs PR test is broken for noe (# 30527), but other than that, things look ready for a review.

@jreback , did you have something like this solution in mind?

AlexKirko commented:
@jreback Merged in master to fix Webs_and_Docs.

jreback commented on pandas/_libs/tslib.pyx:
```diff
@@ -297,16 +297,32 @@ def format_array_from_datetime(ndarray[int64_t] values, object tz=None,
 
 def array_with_unit_to_datetime(ndarray values, object unit,
-                                str errors='coerce'):
+                                str errors='coerce', ndarray mask=None):
```
let's move this after values, and require it

jreback commented on pandas/_libs/tslib.pyx:
```diff
@@ -297,16 +297,32 @@ def format_array_from_datetime(ndarray[int64_t] values, object tz=None,
+    values : ndarray of object
+         Date-like objects to convert
+    unit : object
+         Time unit to use during conversion
```
put these in the same order as the args

jreback commented on pandas/core/tools/datetimes.py:
```diff
@@ -316,8 +317,20 @@ def _convert_listlike_datetimes(
-        arg = getattr(arg, "values", arg)
-        result, tz_parsed = tslib.array_with_unit_to_datetime(arg, unit, errors=errors)
+        arg = getattr(arg, "_values", arg)
+        # GH 30050 pass an ndarray to tslib.array_with_unit_to_datetime
```
blank line before the comment

jreback commented on pandas/core/tools/datetimes.py:
```diff
@@ -316,8 +317,20 @@ def _convert_listlike_datetimes(
+        # because it expects an ndarray argument
+        if isinstance(arg, IntegerArray):
+            # Explicitly pass NaT mask to array_with_unit_to_datetime
+            mask_na = arg.isna()
```
just call this mask

jreback commented on pandas/core/tools/datetimes.py:
```diff
@@ -316,8 +317,20 @@ def _convert_listlike_datetimes(
+            # Explicitly pass NaT mask to array_with_unit_to_datetime
+            mask_na = arg.isna()
+            arg = getattr(arg, "_ndarray_values", arg)
+            result, tz_parsed = tslib.array_with_unit_to_datetime(
```
move this outside the if/else and add ``mask=None`` for the non-IntegerArray case, then you can just call this once

jreback reviewed (changes requested):
lgtm (yes this was the approach was looking for). some small comments, ping on green.

AlexKirko commented:
@jreback Addressed all the comments, please take a look.

TomAugspurger commented on pandas/_libs/tslib.pyx:
```diff
@@ -329,9 +344,13 @@ def array_with_unit_to_datetime(ndarray values, object unit,
+            result, tz = array_to_datetime(values.astype(object), errors=errors)
+        if mask is not None:
+            iresult = result.view('i8')
+            iresult[mask] = NPY_NAT
```
This is going to mutate `result` inplace. Is that what we want?

Given the subtleties here, does it make sense for the *calling* function to handle this, rather than passing a mask down.

jreback replied:
yes i think this is worth it, the calling function is aleady way too complex, and we *already* deal with mask here. the mutating here is a common pattern (we mutate on the i8 and auctally return the m8)

TomAugspurger commented on pandas/core/tools/datetimes.py:
```diff
@@ -316,8 +317,21 @@ def _convert_listlike_datetimes(
+        if isinstance(arg, IntegerArray):
+            # Explicitly pass NaT mask to array_with_unit_to_datetime
+            mask = arg.isna()
+            arg = getattr(arg, "_ndarray_values", arg)
```
Why getattr instead of `arg._ndarray_values`?

TomAugspurger replied:
Once we have IntegerArray.to_numpy, I think this could be something like `arg.to_numpy("i8", fill_value=NPY_NAT)`.

jreback replied:
yes would like to do this exactly (or a way to get data/mask publicly would be ok here as well). but what you are suggesting is nice.

TomAugspurger commented on pandas/tests/indexes/datetimes/test_tools.py:
```diff
@@ -2291,3 +2291,22 @@ def test_should_cache_errors(unique_share, check_count, err_message):
+            np.datetime64("NaT"),
+        ]
+    )
+    tm.assert_series_equal(res, expected)
```
Can you ensure that `ser` isn't mutated inplace?

pep8speaks commented:
Hello @AlexKirko! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`pandas/tests/indexes/datetimes/test_tools.py`](https://github.com/pandas-dev/pandas/blob/1c13f98ddb05155d6322cdf2ff4af8fa02c3dfc9/pandas/tests/indexes/datetimes/test_tools.py):

> [Line 2301:1](https://github.com/pandas-dev/pandas/blob/1c13f98ddb05155d6322cdf2ff4af8fa02c3dfc9/pandas/tests/indexes/datetimes/test_tools.py#L2301): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace

AlexKirko commented:
Added to the test and removed unnecessary `getattr`. Awaiting the decision on where we should implement NaN-handling: in `tslib` or in `datetimes`. I'm for `tslib` (current solution) as we merely expand on the way NaNs are handle and that's, in my opinion, is better than splitting and re-merging arg in `_convert_listlike_datetimes`.

jreback commented on pandas/_libs/tslib.pyx:
```diff
@@ -296,17 +296,33 @@ def format_array_from_datetime(ndarray[int64_t] values, object tz=None,
+    ----------
+    values : ndarray of object
+         Date-like objects to convert
+    mask : ndarray of bool, default None
```
this doesn't default to None; rather it *can* be None

jreback commented on pandas/tests/indexes/datetimes/test_tools.py:
```diff
@@ -2291,3 +2291,25 @@ def test_should_cache_errors(unique_share, check_count, err_message):
         tools.should_cache(arg, unique_share, check_count)
+
+
+def test_intarray_to_datetime():
```
can you say: nullable_integer rather than intarray

jreback reviewed (changes requested):
small comment, pls rebase and ping on green.

AlexKirko committed:
BUG: move nan-handling to tslib and clean up

AlexKirko commented:
@jreback rebased to updated master, squashed tiny commits, made the needed changes and commented on green

TomAugspurger reviewed (approved):
Thanks @AlexKirko, this looks good. Merging later today, in case @jreback wants to have one more look.

TomAugspurger commented:
Thanks @AlexKirko!

AlexKirko commented:
@jreback @TomAugspurger Thank you! Learned a lot about the process: it should help a lot with the next PR.
